pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+
+function debug(...args: unknown[]) {
+    if (window.debugging == 'all' || window.debugging?.includes('upload'))
+        console.debug('debug[upload]', ...args);
+}
+
+class UploadError extends Error {
+    name = 'UploadError';
+    constructor(problem: string, message: cockpit.JsonObject) {
+        // XXX message.message is untranslated.  Should we just always use cockpit.message(problem)?
+        super(typeof message.message === 'string' ? message.message : cockpit.message(problem));
+    }
+}
+
+/* XXX: these values are ~a decade old and were more or less copied from ssh
+ * (where they were last changed in 2007).  We might test alternatives...
+ */
+const BLOCK_SIZE = 16 << 10; // 16kiB
+const FLOW_WINDOW = 2 << 20; // 2MiB
+
+export async function upload(
+    destination: string,
+    stream: ReadableStream,
+    progress?: (bytes_sent: number) => void,
+    signal?: AbortSignal,
+    options?: cockpit.JsonObject
+) {
+    /* While the upload is running, the function is essentially always waiting
+     * inside of a Promise resolved by notify().  Anything that might cause
+     * forward progress (ack messages, close messages, cancellation) will call
+     * notify().  It's up to the main loop body to re-evaluate and respond.
+     */
+    let outstanding = 0; // for flow control
+    let delivered = 0; // for progress reporting
+    let close_message = null as cockpit.JsonObject | null;
+    let notify = () => { };
+
+    if (signal) {
+        signal.throwIfAborted(); // early exit
+        signal.addEventListener('abort', () => { notify() });
```

Comment:
Make this `() => notify?.()` to support an unset handler?

But moreover, "abort" sounds like a terminal action for the promise, so it should get rejected, not notified?

I find `notify` a really confusing name. It immediately reminds me of `cockpit.defer.notify`, which is a thing, but obsolete. It also raises the questions "this should reject the promise, not notify it", and "why doesn't this receive arguments"? 

I think what this *really* wants to be is `let upload_promise_resolve = null` or similar, to make sure that this is a Promise resolve callback, not a cockpit.defer.notify() one.

Reply:
After the force-push: Is it possible to move this inside of the `new Promise` block, so that you don't have to do this dance at all, and can use the `resolve` handler as-is?

Reply:
This is named `notify` for the `notify()` method found on condition variables (which is, in effect, what this thing is — a way to block until some state has changed).

Let's chat about the `?.()` idea.  It strikes me as a bit ugly, but I guess you'd say that having the initial `() => { }` value strikes you as a bit ugly... :)

Reply:
Also, about the abort: a previous version of this patch indeed handled abort by rejecting the promise, which *usually* worked (and indeed, the tests always passed) but there was a theoretical possibility that the abort would come in, either: 
 - while we were `await`ing the file to be read from the disk; or
 - after an `ack` message had already been delivered in the main loop iteration, but before we could return to our main function body.
In this case, the promise would have already been resolved, the rejection a no-op, and the cancellation lost.  That's why I check it in the loop body now — and it also lets the `AbortSignal` itself raise the exception, without us getting involved.

Reply:
I'd prefer creating the `Promise` earlier and drop all the conditionals and reassigning (that's what I meant with "drop this dance"). Then you have the `resolve` and reject` handlers everywhere you need them.

Reply:
See above, let's not restructure the code again. I'd just *really* like to rename this from `notify` to `resolve` or naming the promise "channel_event" and then this `channel_event_resolve`. notify has such a big history with both cockpit.defer and also e.g. the D-Bus and metrics channels that it's too confusing.